pkg/store/bucket.go

```diff
@@ -2547,6 +2575,47 @@ func (r *bucketIndexReader) fetchPostings(ctx context.Context, keys []labels.Lab
 	return output, closeFns, g.Wait()
 }
 
+func (r *bucketIndexReader) decodeCachedPostings(b []byte) (index.Postings, []func(), error) {
+	// Even if this instance is not using compression, there may be compressed
+	// entries in the cache written by other stores.
+	var (
+		l        index.Postings
+		err      error
+		closeFns []func()
+	)
+	if isDiffVarintSnappyEncodedPostings(b) || isDiffVarintSnappyStreamedEncodedPostings(b) {
+		s := time.Now()
+		clPostings, err := decodePostings(b)
+		r.stats.cachedPostingsDecompressions += 1
+		r.stats.CachedPostingsDecompressionTimeSum += time.Since(s)
+		if err != nil {
+			r.stats.cachedPostingsDecompressionErrors += 1
+		} else {
+			closeFns = append(closeFns, clPostings.close)
+			l = clPostings
+		}
+	} else {
+		_, l, err = r.dec.Postings(b)
+	}
+	return l, closeFns, err
+}
+
+func (r *bucketIndexReader) encodePostingsToCache(p index.Postings, length int) ([]byte, time.Duration, int, int) {
```

Comment:
Small nit but maybe instead of returning multiple `int`s we could pass `r.stats.cachedPostingsCompressionErrors` and `r.stats.CachedPostingsCompressedSizeSum` directly? Perhaps even the whole `r.stats`? I think this would make the code clearer.

Reply:
Yeah that makes sense and I wanted to do the same at the beginning.
But since `encodePostingsToCache` will be called concurrently, passing r.stats means we need to have lock inside https://github.com/thanos-io/thanos/pull/6420/files#diff-3e2896fafa6ff73509c77df2c4389b68828e02575bb4fb78b6c34bcfb922a7ceR2554.

But not all situations need to hold the lock so I want to simplify the logic and the caller can decide how to update the stats and add lock. WDYT?

Reply:
Ok, that makes sense. Maybe it is logical to convert those fields to atomic types to avoid holding a lock in a separate PR so that it could simplify this function's return values.